test/e2e/fqdn_dns_cache_test.go

```diff
@@ -0,0 +1,391 @@
+
+	if err != nil {
+		return "", err
+	}
+
+	return resolvedIP.String(), nil
+}
+
+// createAndDeployCustomDnsConfigMap creates and deploys the custom DNS ConfigMap with the specified IP and FQDN.
+func createAndDeployCustomDnsConfigMap(t *testing.T, data *TestData, ipv4 string, testFQDN string) (*corev1.ConfigMap, error) {
+	const customDnsConfigName = "custom-dns-config"
+	dnsConfigData := createDnsConfig(t, ipv4, testFQDN)
+	configMap := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      customDnsConfigName,
+			Namespace: data.testNamespace,
+		},
+		Data: dnsConfigData,
+	}
+	return data.CreateConfigMap(configMap)
+
+}
+
+// updateCustomDnsConfigMap updates the existing custom DNS ConfigMap with a new IP for the specified FQDN.
+func updateCustomDnsConfigMap(t *testing.T, data *TestData, ipv4 string, testFullyQualifiedDomainName string, customDnsConfigMap *corev1.ConfigMap) error {
+	customDnsConfigMap.Data = createDnsConfig(t, ipv4, testFullyQualifiedDomainName)
+	return data.UpdateConfigMap(customDnsConfigMap)
+}
+
+// setCustomDnsServerAddressInAntrea sets or resets the custom DNS server IP address in Antrea configMap.
+func setCustomDnsServerAddressInAntrea(t *testing.T, data *TestData, teardown bool) string {
+	var dnsServiceIP string
+
+	// Create or reset DNS service based on teardown flag
+	if !teardown {
+		var err error
+		dnsServiceIP, err = createCustomDnsService(data)
+		require.NoError(t, err, "Error when creating custom DNS service: %v", err)
+	}
+
+	err := updateAntreaConfigMap(data, dnsServiceIP)
+	require.NoError(t, err, "Error when updating Antrea configmap with custom dns ip : %v", err)
+
+	if teardown {
+		t.Logf("removing dns server IP from antrea agent during teardown")
+		return ""
+	}
+	t.Logf("Set DNS server IP in Antrea : %s", dnsServiceIP)
+	return dnsServiceIP
+}
+
+// updateAntreaConfigMap updates the Antrea configuration with the specified DNS server IP.
+func updateAntreaConfigMap(data *TestData, dnsServiceIP string) error {
+	agentChanges := func(config *agentconfig.AgentConfig) {
+		config.DNSServerOverride = dnsServiceIP
+	}
+	return data.mutateAntreaConfigMap(nil, agentChanges, false, true)
+}
+
+// createCustomDnsService creates the custom DNS service and returns its ClusterIP as ipv4 address.
+func createCustomDnsService(data *TestData) (string, error) {
+	const dnsPort = 53
+	customDnsService, err := data.CreateServiceWithAnnotations("custom-dns-service", data.testNamespace, dnsPort,
+		dnsPort, corev1.ProtocolUDP, map[string]string{"app": "custom-dns"}, false,
+		false, corev1.ServiceTypeClusterIP, ptr.To[corev1.IPFamily](corev1.IPv4Protocol), map[string]string{})
+	if err != nil {
+		return "", err
+	}
+	return customDnsService.Spec.ClusterIP, nil
+
+}
+
+// createDnsConfig generates a DNS configuration for the specified IP address and domain name.
+func createDnsConfig(t *testing.T, ipv4Address, domainName string) map[string]string {
```

Comment:
there is nothing specific to "ipv4" in the function implementation.
Just use something like `func createDnsConfig(t *testing.T, hosts map[string]string) map[string]string`, where `hosts` maps IP addresses to host names.